File: sdks/typescript/src/runtime/runtime.ts
import { decodeFrame } from './messages.js';
import type { InboundFrame, Invocation, SdkPayload } from './messages.js';
import { cleanupAll, executeWithCleanup } from './cleanup.js';
import {
    FailureCodes,
    FrameKinds,
    ProtocolLimits,
    ProtocolVersions,
    SdkOperations,
} from '../protocol.js';
import { StreamBuffer } from './stream-buffer.js';
import { SourceSession } from '../source.js';
import type { SourceHandler } from '../source.js';
import { runConcurrent } from './concurrent.js';
import { AsyncLocalStorage } from 'node:async_hooks';
import { Console } from 'node:console';
import { createConnection, Socket } from 'node:net';
import { createInterface } from 'node:readline';
import { randomUUID } from 'node:crypto';
import type { Readable, Writable } from 'node:stream';

import { SessionContext, SessionCleanupError } from '../session.js';
import type { Handler, Generator } from './handlers.js';
type InvocationScope = { request?: Invocation; active: boolean };
class StreamClosedError extends Error {}
const scopes = new AsyncLocalStorage<InvocationScope>();
const encode = (value: unknown) => {
    const text = JSON.stringify(value);
    if (text === undefined) throw new TypeError('Result is not JSON');
    return text;
};

export class Runtime {
    private reader!: AsyncIterator<string>;
    private output!: Writable;
    private socket?: Socket;
    private source = new SourceSession();
    private streamBuffer = new StreamBuffer(encode);
    private iterator?: AsyncIterator<unknown>;
    private context?: SessionContext;
    private abort?: AbortController;
    private streamId?: string;
    private callbackId = 0;
    private streamScope?: InvocationScope;
    private exchangeActive = false;
    private bufferedStreamItems = false;
    private closingStream = false;
    private exchangeWaiters: (() => void)[] = [];
    private callbackTail: Promise<void> = Promise.resolve();
    constructor(
        private functions: Map<string, Handler>,
        private streams: Map<string, Generator>,
        private sources: Map<string, SourceHandler>,
        private pluginVersion: string,
        private concurrentCalls: boolean,
    ) {}
    private async read(): Promise<InboundFrame | undefined> {
        const line = await this.reader.next();
        if (line.done) return undefined;
        if (Buffer.byteLength(line.value) > ProtocolLimits.FrameBytes)
            throw new Error('Frame limit');
        return decodeFrame(JSON.parse(line.value));
    }
    private async send(value: unknown): Promise<void> {
        const line = encode(value);
        if (Buffer.byteLength(line) > ProtocolLimits.FrameBytes) throw new Error('Frame limit');
        await new Promise<void>((resolve, reject) =>
            this.output.write(line + '\n', (error) => (error ? reject(error) : resolve())),
        );
    }
    private async callback<T>(operation: string, payload: unknown): Promise<T> {
        const scope = scopes.getStore();
        const streamCallback = scope === this.streamScope;
        if (streamCallback && (!this.exchangeActive || this.bufferedStreamItems)) {
            await new Promise<void>((resolve) => this.exchangeWaiters.push(resolve));
        }
        if (streamCallback && this.closingStream) throw new StreamClosedError('Stream closed');
        let release!: () => void;
        const previous = this.callbackTail;
        this.callbackTail = new Promise<void>((resolve) => (release = resolve));
        await previous;
        try {
            if (!scope?.active || !scope.request) throw new Error('Expired invocation');
            const callbackId = String(++this.callbackId),
                id = scope.request.id;
            await this.send({ type: FrameKinds.Callback, id, callbackId, operation, payload });
            const reply = await this.read();
            if (
                reply?.type !== 'callback-result' ||
                reply.id !== id ||
                reply.callbackId !== callbackId
            )
                throw new Error('Callback identity');
            if (reply.error != null || reply.success === false)
                throw new Error('Host callback failed');
            return reply.value as T;
        } finally {
            release();
        }
    }
    private async close(): Promise<void> {
        const iterator = this.iterator;
        this.closingStream = true;
        this.source.release();
        const advance = this.streamBuffer.reset();
        this.streamScope = undefined;
        this.bufferedStreamItems = false;
        this.iterator = undefined;
        this.streamId = undefined;
        this.abort?.abort();
        this.abort = undefined;
        await cleanupAll(
            async () => {
                try {
                    if (advance) await advance;
                } catch (error) {
                    if (!(error instanceof StreamClosedError)) throw error;
                }
            },
            async () => {
                if (iterator?.return) await iterator.return();
            },
            () => this.completeContext(),
        );
    }
    private async completeContext(): Promise<void> {
        const context = this.context;
        this.context = undefined;
        await context?.complete();
    }
    private async dispatch(request: Invocation): Promise<unknown> {
        const { operation, payload } = request;
        if (operation === SdkOperations.SourceRead || operation === SdkOperations.SourceClose) {
            return await this.dispatchSource(operation, payload);
        }
        if (operation === SdkOperations.Next || operation === SdkOperations.Close) {
            return await this.dispatchStream(operation, payload);
        }
        if (this.iterator || this.source.identity) throw new Error('Session already active');
        const context = (this.context = new SessionContext(
            request.context.tenant,
            request.context.configuration,
            <T>(name: string, input: unknown) => this.callback<T>(name, input),
        ));
        const abort = new AbortController();
        if (operation === SdkOperations.SourceOpen) {
            this.abort = abort;
            return await this.source.open(
                this.sources.get(payload.operation ?? '')!,
                payload.input,
                context,
                abort.signal,
            );
        }
        if (operation === SdkOperations.Call) {
            return await executeWithCleanup(
                () =>
                    this.functions.get(payload.operation ?? '')!(
                        payload.input,
                        context,
                        abort.signal,
                    ),
                () =>
                    cleanupAll(
                        async () => {
                            abort.abort();
                        },
                        () => this.completeContext(),
                    ),
            );
        }
        if (operation !== SdkOperations.Start) throw new Error('Unknown SDK operation');
        this.abort = abort;
        this.closingStream = false;
        this.streamScope = scopes.getStore();
        this.iterator = this.streams.get(payload.operation ?? '')!(
            payload.input,
            context,
            abort.signal,
        )[Symbol.asyncIterator]();
        this.streamId = randomUUID();
        return { stream: this.streamId };
    }
    private async connect(): Promise<void> {
        let input: Readable = process.stdin;
        this.output = process.stdout;
        if (process.env.WEAVEPORT_SOCKET) {
            const socket = createConnection(process.env.WEAVEPORT_SOCKET);
            await new Promise<void>((resolve, reject) => {
                socket.once('connect', resolve);
                socket.once('error', reject);
            });
            this.socket = socket;
            input = socket;
            this.output = socket;
        }
        globalThis.console = new Console({
            stdout: process.stderr,
            stderr: process.stderr,
        });
        this.reader = createInterface({ input, crlfDelay: Infinity })[Symbol.asyncIterator]();
    }

    async run(): Promise<void> {
        await this.connect();
        await this.send({
            type: FrameKinds.Ready,
            protocol: this.concurrentCalls
                ? ProtocolVersions.Concurrent
                : ProtocolVersions.Exclusive,
            pluginVersion: this.pluginVersion,
            sessionCleanup: ProtocolVersions.SessionCleanup,
            ...(this.concurrentCalls ? { concurrentCalls: ProtocolVersions.ConcurrentCalls } : {}),
        });
        if (this.concurrentCalls) {
            try {
                await runConcurrent(
                    () => this.read(),
                    (value) => this.send(value),
                    this.functions,
                );
            } finally {
                this.socket?.destroy();
            }
            return;
        }
        try {
            let request: InboundFrame | undefined;
            while ((request = await this.read()) !== undefined) {
                if (request.type !== FrameKinds.Invoke) throw new Error('Expected invocation');
                const invocation = request;
                const scope = this.streamScope ?? { request: invocation, active: true };
                scope.request = invocation;
                scope.active = true;
                this.exchangeActive = true;
                this.resumeCallbacks();
                await this.exchange(invocation, scope);
                request = undefined;
            }
        } finally {
            await this.close();
            this.socket?.destroy();
        }
    }

    private async exchange(invocation: Invocation, scope: InvocationScope): Promise<void> {
        await scopes.run(scope, async () => {
            try {
                const value = await this.dispatch(invocation);
                this.exchangeActive = false;
                scope.active = false;
                await this.callbackTail;
                await this.send({
                    type: FrameKinds.Result,
                    id: invocation.id,
                    value,
                    reusable: !this.context && !this.iterator,
                });
            } catch (error) {
                let cleanupFailed = error instanceof SessionCleanupError;
                try {
                    await this.close();
                } catch {
                    cleanupFailed = true;
                }
                await this.send({
                    type: FrameKinds.Error,
                    id: invocation.id,
                    code: cleanupFailed ? FailureCodes.CleanupError : FailureCodes.SdkError,
                    primaryCode:
                        error instanceof SessionCleanupError && !error.hasExecutionFailure
                            ? FailureCodes.CleanupError
                            : FailureCodes.SdkError,
                    cleanupFailed,
                });
            } finally {
                this.exchangeActive = false;
                scope.active = false;
                scope.request = undefined;
            }
        });
    }
    private async dispatchSource(operation: string, payload: SdkPayload): Promise<unknown> {
        if (!this.source.identity || payload.source !== this.source.identity)
            throw new Error('Source ownership');
        if (operation === SdkOperations.SourceClose) {
            await this.close();
            return {};
        }
        const result = await this.source.read(payload);
        if (result.done) await this.close();
        return result;
    }
    private async dispatchStream(operation: string, payload: SdkPayload): Promise<unknown> {
        if (!this.streamId || payload.stream !== this.streamId) throw new Error('Stream ownership');
        if (operation === SdkOperations.Close) {
            await this.close();
            return {};
        }
        this.bufferedStreamItems = false;
        const batch = await this.streamBuffer.read(this.iterator!, () => {
            this.bufferedStreamItems = true;
        });
        if (batch.done) await this.close();
        return batch;
    }
    private resumeCallbacks(): void {
        for (const resume of this.exchangeWaiters.splice(0)) resume();
    }
}

File: sdks/python/weaveport_sdk/runtime.py
"""Exclusive protocol runtime; one invocation scope owns callback authority."""

import asyncio
import contextvars
import json
import os
import socket
import sys
import uuid

from .protocol import (
    ProtocolVersions,
    FailureCodes,
    FrameKinds,
    ProtocolLimits,
    SdkOperations,
)

from .cleanup import cleanup_all, execute_with_cleanup
from .context import PluginContext, SessionCleanupError
from .source import SourceSession
from .stream import LiveStream

_CHANNEL_BUFFER_BYTES = 65536

_SCOPE = contextvars.ContextVar("weaveport_invocation", default=None)


def _encode(value):
    return json.dumps(
        value, ensure_ascii=True, allow_nan=False, separators=(",", ":")
    ).encode("utf-8")


class Runtime:
    def __init__(self, app):
        self.app = app
        self.source = SourceSession()
        self.context = None
        self.iterator = None
        self.stream_id = None
        self.callback_id = 0
        self.callback_lock = asyncio.Lock()
        self.exchange_ready = asyncio.Event()
        self.stream_scope = None
        self.live_stream = None
        self.channel = None
        if endpoint := os.getenv("WEAVEPORT_SOCKET"):
            self.channel = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            self.channel.connect(endpoint)
            self.reader = self.channel.makefile("rb")
            self.writer = self.channel.makefile("wb", buffering=_CHANNEL_BUFFER_BYTES)
        else:
            self.reader = sys.stdin.buffer
            self.writer = sys.stdout.buffer
        sys.stdout = sys.stderr

    async def read(self):
        line = await asyncio.to_thread(
            self.reader.readline, ProtocolLimits.FrameBytes + 2
        )
        if not line:
            return None
        if len(line) > ProtocolLimits.FrameBytes + 1 or not line.endswith(b"\n"):
            raise ValueError("Frame limit")
        return json.loads(line)

    async def send(self, value):
        data = _encode(value)
        if len(data) > ProtocolLimits.FrameBytes:
            raise ValueError("Frame limit")
        self.writer.write(data + b"\n")
        self.writer.flush()

    async def callback(self, operation, value):
        scope = _SCOPE.get()
        if scope is self.stream_scope:
            await self.wait_for_stream_exchange()
        async with self.callback_lock:
            if scope is None or not scope["active"]:
                raise RuntimeError("Expired invocation")
            self.callback_id += 1
            callback_id = str(self.callback_id)
            request_id = scope["request"]["id"]
            await self.send(
                dict(
                    type=FrameKinds.Callback,
                    id=request_id,
                    callbackId=callback_id,
                    operation=operation,
                    payload=value,
                )
            )
            reply = await self.read()
            if (
                reply is None
                or reply.get("type") != FrameKinds.CallbackResult
                or reply.get("id") != request_id
                or reply.get("callbackId") != callback_id
            ):
                raise ValueError("Callback identity")
            if reply.get("error") is not None or reply.get("success") is False:
                raise RuntimeError("Host callback failed")
            return reply["value"]

    async def close(self):
        iterator, self.iterator = self.iterator, None
        live_stream, self.live_stream = self.live_stream, None
        self.stream_scope = None
        self.stream_id = None
        self.source.release()
        actions = []
        if live_stream is not None:
            actions.append(live_stream.close)
        if iterator is not None:
            actions.append(iterator.aclose)
        actions.append(self.complete_context)
        await cleanup_all(*actions)

    async def complete_context(self):
        context, self.context = self.context, None
        if context is not None:
            await context._close()

    async def dispatch(self, request):
        operation, payload = request["operation"], request["payload"]
        if operation in (SdkOperations.SourceRead, SdkOperations.SourceClose):
            return await self.dispatch_source(operation, payload)
        if operation in (SdkOperations.Next, SdkOperations.Close):
            return await self.dispatch_stream(operation, payload)
        if self.iterator is not None or self.source.identity is not None:
            raise RuntimeError("Session already active")
        context = self.context = PluginContext(
            request["context"]["tenant"], request["context"]["configuration"], self
        )
        if operation == SdkOperations.SourceOpen:
            return await self.source.open(
                self.app._sources[payload["operation"]], payload["input"], context
            )
        if operation == SdkOperations.Call:
            return await execute_with_cleanup(
                lambda: self.app._functions[payload["operation"]](
                    payload["input"], context
                ),
                self.complete_context,
            )
        if operation != SdkOperations.Start:
            raise ValueError("Unknown SDK operation")
        self.stream_scope = _SCOPE.get()
        self.iterator = self.app._streams[payload["operation"]](
            payload["input"], context
        ).__aiter__()
        self.live_stream = LiveStream(self.iterator, _encode)
        self.stream_id = uuid.uuid4().hex
        return dict(stream=self.stream_id)

    async def run(self):
        ready = dict(
            type=FrameKinds.Ready,
            protocol=(
                ProtocolVersions.Concurrent
                if self.app._concurrent_calls
                else ProtocolVersions.Exclusive
            ),
            pluginVersion=self.app._plugin_version,
            sessionCleanup=ProtocolVersions.SessionCleanup,
        )
        if self.app._concurrent_calls:
            ready["concurrentCalls"] = ProtocolVersions.ConcurrentCalls
        await self.send(ready)
        if self.app._concurrent_calls:
            from .concurrent import run_concurrent

            try:
                await run_concurrent(self, PluginContext, _SCOPE)
            finally:
                self.close_channel()
            return
        try:
            while (request := await self.read()) is not None:
                await self.exchange(request)
                request = None
        finally:
            await self.close()
            if self.channel:
                self.reader.close()
                self.writer.close()
                self.channel.close()

    async def exchange(self, request):
        """Keep callback authority active only for this host exchange."""
        scope = self.stream_scope or dict(request=request, active=True)
        scope.update(request=request, active=True)
        self.exchange_ready.set()
        handle = _SCOPE.set(scope)
        try:
            value = await self.dispatch(request)
            self.exchange_ready.clear()
            scope["active"] = False
            async with self.callback_lock:
                pass
            await self.send(
                dict(
                    type=FrameKinds.Result,
                    id=request["id"],
                    value=value,
                    reusable=self.context is None and self.iterator is None,
                )
            )
        except Exception as error:
            cleanup_failed = isinstance(error, SessionCleanupError)
            try:
                await self.close()
            except Exception:
                cleanup_failed = True
            await self.send(
                dict(
                    type=FrameKinds.Error,
                    id=request["id"],
                    code=(
                        FailureCodes.CleanupError
                        if cleanup_failed
                        else FailureCodes.SdkError
                    ),
                    primaryCode=(
                        FailureCodes.CleanupError
                        if isinstance(error, SessionCleanupError)
                        and not getattr(error, "has_execution_failure", False)
                        else FailureCodes.SdkError
                    ),
                    cleanupFailed=cleanup_failed,
                )
            )
        finally:
            self.exchange_ready.clear()
            scope["active"] = False
            _SCOPE.reset(handle)
            scope["request"] = None
            value = None

    async def dispatch_source(self, operation, payload):
        if self.source.identity is None or payload["source"] != self.source.identity:
            raise ValueError("Source ownership")
        if operation == SdkOperations.SourceClose:
            await self.close()
            return {}
        result = await self.source.read(payload)
        if result["done"]:
            await self.close()
        return result

    async def dispatch_stream(self, operation, payload):
        if self.stream_id is None or payload["stream"] != self.stream_id:
            raise ValueError("Stream ownership")
        if operation == SdkOperations.Close:
            await self.close()
            return {}
        result = await self.live_stream.next_batch()
        if result["done"]:
            await self.close()
        return result

    async def wait_for_stream_exchange(self):
        if self.live_stream is not None:
            await self.live_stream.wait_for_delivery()
        await self.exchange_ready.wait()

    def close_channel(self):
        if self.channel:
            self.channel.close()
